tensorflow_privacy/privacy/analysis/rdp_accountant.py

```diff
@@ -313,6 +379,128 @@ def compute_rdp(q, noise_multiplier, steps, orders):
+  """Compute RDP of the Sampled Gaussian mechanism at order alpha.
+  Args:
+    q: The sampling proportion =  m / n.  Assume m is an integer <= n.
+    sigma: The std of the additive Gaussian noise.
+    alpha: The order at which RDP is computed.
+  Returns:
+    RDP at alpha, can be np.inf.
+  """
+
+  assert (q <= 1) and (q >= 0) and (alpha >= 1)
+
+  if q == 0:
+    return 0
+
+  if q == 1.:
+    return alpha / (2 * sigma**2)
+
+  if np.isinf(alpha):
+    return np.inf
+
+
+
+  if isinstance(alpha, six.integer_types):
+    return _compute_rdp_sample_without_replacement_int(q, sigma, alpha) / (alpha - 1)
+  else:
+    # When alpha not an integer, we apply Corollary 10 of [WBK19] to interpolate the
+    # CGF and obtain an upper bound
+    alpha_f = math.floor(alpha)
+    alpha_c = math.ceil(alpha)
+
+    x = _compute_rdp_sample_without_replacement_int(q, sigma, alpha_f)
+    y = _compute_rdp_sample_without_replacement_int(q, sigma, alpha_c)
+    t = alpha - alpha_f
+    return ((1-t) * x + t * y) / (alpha-1)
+
+def _compute_rdp_sample_without_replacement_int(q, sigma, alpha):
+  """Compute log(A_alpha) for integer alpha. 0 < q < 1, under subsampling without replacement.
+  when alpha is smaller than max_alpha, compute the bound Theorem 27 exactly, else compute the bound with stirling approximation
+    Args:
+    q:  The sampling proportion =  m / n.  Assume m is an integer <= n.
+    sigma: The std of the additive Gaussian noise.
+    alpha: The order at which RDP is computed.
+  Returns:
+    RDP at alpha, can be np.inf.
+  """
+
+  max_alpha = 100
+  assert isinstance(alpha, six.integer_types)
+
+  if np.isinf(alpha):
+    return np.inf
+  elif alpha==1:
+    return 0
+
+  def cgf(x):
+    # Return rdp(x+1)*x, the rdp of Gaussian mechanism is alpha/(2*sigma**2)
+    return x*1.0*(x+1)/(2.0*sigma**2)
+
+  def func(x):
+    # Return the rdp of Gaussian mechanism
+    return 1.0*(x)/(2.0*sigma**2)
+
+  # We need forward differences of exp(cgf)
+  # The following line is the numerically stable way of implementing it.
+  # The output is in polar form with logarithmic magnitude
+  deltas, signs_deltas = _get_forward_diffs(cgf, alpha)
+
+
+  # Initialize with 1 in the log space.
+  log_a = 0
+  if alpha <= max_alpha:
+    # Compute the bound exactly requires book keeping of O(alpha**2)
+
+    for i in range(2, alpha+1):
+      if i == 2:
+        s = 2 * np.log(q) + _log_comb(alpha, 2)  + np.minimum(np.log(4) + func(2.0) + np.log(1 - np.exp(-func(2.0))),func(2.0) + np.log(2))
+      elif i > 2:
+        s = np.minimum(np.log(4) + 0.5*deltas[int(2*np.floor(i/2.0))-1]+ 0.5*deltas[int(2*np.ceil(i/2.0))-1],np.log(2)+ cgf(i - 1)) \
+                                                 + i * np.log(q) +_log_comb(alpha, i)
+      log_a = _log_add(log_a,s)
+    return float(log_a)
+  else:
+    # Compute the bound with stirling approximation. Everything is O(x) now.
+    for i in range(2, alpha + 1):
```

Comment:
If alpha is very large (millions, say) this will take forever. Is there anything we can do in that case?

Reply:
The alpha list is given as the input (in orders) and this part only costs O(alpha) complexity. I think in Poisson subsample we also have a O(alpha) complexity there? https://github.com/tensorflow/privacy/blob/bd69c70965b3cf528b457081c98f43a6a160aa4c/tensorflow_privacy/privacy/analysis/rdp_accountant.py#L100

Reply:
That's true, just considering whether there could be any optimization.